numpy/core/tests/test_multiarray.py

```diff
@@ -6021,7 +6022,28 @@ def test_dot_equivalent(self, args):
 
         r3 = np.matmul(args[0].copy(), args[1].copy())
         assert_equal(r1, r3)
-        
+    
+    def test_matmul_object(self):
+        import fractions
+
+        f = np.vectorize(fractions.Fraction)
+        def random_ints():
+            return np.random.randint(1000, size=(10, 3, 3))
+        M1 = f(random_ints(), random_ints()) 
+        M2 = f(random_ints(), random_ints())
+
+        M3 = self.matmul(M1, M2)
+
+        [N1, N2, N3] = [a.astype(float) for a in [M1, M2, M3]]
+
+        assert_allclose(N3, self.matmul(N1, N2))
+
+    def test_matmul_object_type_scalar(self):
+        from fractions import Fraction as F
+        v = np.array([F(2,3), F(5,7)])
+        res = self.matmul(v, v)
+        assert_(type(res) is F)
+
```

Comment:
Can you make sure you have a test for something like:

```
np.matmul(np.empty((3, 0), dtype=object), np.empty((0, 3), dtype=object))
```
It probably should return an array filled with `False` (different then `np.dot`, but `np.dot` seems broken currently).

It would also be good to test both error path. The multiply path is simple, since you can just put a `None` or so inside. The `add` error path we probably have to create a dummy object:
```
class multiply_not_add(object):
    def __multiply__(self, other):
        return self
    def __add__(self, other):
        raise ValueError("this object cannot be added")
```
should work.

Reply:
OK, just checked `np.sum(np.array([], dtype=object))` returns `0` Integer, so I guess we could go that way. (I thought that one was `False`)

Reply:
I think it should be full of the integer `0` which matches `np.sum` - not `False`

Reply:
I agree it should be `0`.